feat(types-registry): add Phase 1.1 OpenSpec proposal

alizid10 opened the pull request:
Scaffold types-registry module with SDK contracts, in-memory storage,
and gts-rust integration for GTS operations. Includes proposal.md,
design.md, tasks.md, and spec deltas.

alizid10 committed:
feat(types-registry): add Phase 1.1 OpenSpec proposal

Add OpenSpec change proposal for types-registry module Phase 1.1.

This proposal establishes the foundational GTS (Global Type System) registry
with SDK contracts and in-memory storage. Key components:

- TypesRegistryApi trait with register_entity, list_entities, get_entity
- GtsEntity model with deterministic UUID mapping from GTS identifiers
- In-memory storage using DashMap for thread-safe concurrent access
- Integration with gts-rust library for 11 standard GTS operations

Phase 1.1 scope:
- SDK crate (types-registry-sdk) with public API
- Module implementation with DDD-light pattern
- In-memory persistence (database deferred to Phase 1.3)

Signed-off-by: Anton S <aliziddev@gmail.com>

alizid10 committed:
Update design, use users_info as an example

Signed-off-by: Anton S <aliziddev@gmail.com>

MikeFalcon77 commented on modules/types-registry/openspec/changes/add-types-registry-phase1/specs/types-registry/spec.md:
```diff
@@ -0,0 +1,370 @@
+- Support concurrent read/write access
+- Maintain entity ordering by registration time
+- Support efficient lookup by GTS ID
+- Support efficient pattern-based queries
```
Accroding to InMemoryStorage representation (above) how we are going to search by pattern effectively? InMemoryStorage contains key/value structures only.

alizid10 replied:
Addressed in the code

MikeFalcon77 commented on modules/types-registry/openspec/changes/add-types-registry-phase1/design.md:
```diff
@@ -0,0 +1,208 @@
+InMemoryStorage {
+    entities: DashMap<String, GtsEntity>,  // Key: GTS ID string
+    by_uuid: DashMap<Uuid, String>,        // UUID → GTS ID lookup
+    by_vendor: DashMap<String, HashSet<String>>,  // Vendor → GTS IDs index
```
Dashmap is thread-safe but HashSet is not thread-safe. Need to double-check if it's ok for us.

Artifizer commented on modules/types-registry/openspec/changes/add-types-registry-phase1/specs/types-registry/spec.md:
```diff
@@ -0,0 +1,370 @@
+- **GIVEN** a valid GTS type identifier `gts.acme.core.events.user_created.v1~`
+- **AND** a valid JSON Schema as content
+- **WHEN** the user calls `register` with the entity data
+- **THEN** the system returns the registered entity with generated UUID
```
generated deterministic UUID (Using OP#5), right?

Artifizer commented on modules/types-registry/openspec/changes/add-types-registry-phase1/specs/types-registry/spec.md:
```diff
@@ -0,0 +1,370 @@
+Listing SHALL support:
+- Filtering by GTS ID pattern using OP#4 (ID Pattern Matching) with wildcards
+- Filtering by entity kind (Type, Instance, or both)
+- Filtering by vendor, package, namespace, or type components
```
it's better to explain how it differs from wildcard, if I filter entities by vendor `globex`, than I should get something like this, that I can't get using wildcard:

```
gts.a.b.c.d.v1~globex.app.x.y.v1
gts.k.l.m.n.v1~globex.app.a.b.v1
gts.acme.x.y.z.v1~acme.a.b.c.v1~globex.app.a.b.v1
```

Make it makes sense to add as GTS OP#12 or so as separate PR

Artifizer commented on modules/types-registry/openspec/changes/add-types-registry-phase1/specs/types-registry/spec.md:
```diff
@@ -0,0 +1,370 @@
+
+The system SHALL generate deterministic UUIDs from GTS identifiers.
+
+UUID generation SHALL:
```
I would simply say it should invoke appropriate gts-rust method, not to re-invent it's own semantic

Artifizer commented on modules/types-registry/openspec/changes/add-types-registry-phase1/specs/types-registry/spec.md:
```diff
@@ -0,0 +1,370 @@
+UUID generation SHALL:
+- Produce the same UUID for the same GTS ID (deterministic)
+- Use UUID v5 with a GTS-specific namespace
+- Support scoping by major version (same UUID for v1.0 and v1.5)
```
The gts-rust works differently, it converts the full name to UUID, not just major part, otherwise all the minor versions would have the same UUID that is wrong

You can see all the tests in the spec:
https://github.com/GlobalTypeSystem/gts-spec/tree/main/tests

alizid10 committed:
integrate gts-rust validations, simplify API design

Signed-off-by: Anton S <aliziddev@gmail.com>

alizid10 committed:
split Phase 1.1 into SDK and Module OpenSpec changes

Signed-off-by: Anton S <aliziddev@gmail.com>

alizid10 committed:
feat(types-registry-sdk): implement SDK crate with gts-rust integration

Add types-registry-sdk with TypesRegistryApi trait, GtsEntity/ListQuery models,
and TypesRegistryError. Integrates gts-rust submodule for GtsIdSegment parsing.

Signed-off-by: Anton S <aliziddev@gmail.com>

alizid10 committed:
checkout git submodules in all CI jobs

Signed-off-by: Anton S <aliziddev@gmail.com>

alizid10 committed:
chore: update gts-rust submodule with clippy lint overrides

Signed-off-by: Anton S <aliziddev@gmail.com>

MikeFalcon77 commented on modules/types-registry/types-registry-sdk/src/api.rs:
```diff
@@ -0,0 +1,80 @@
+///
+/// All methods require a `SecurityCtx` for proper authorization and access control.
+#[async_trait]
+pub trait TypesRegistryApi: Send + Sync {
```
PROPOSAL: Starting with this PR, I suggest renaming all these clients from XXXXApi to XXXClient, as the Api suffix is confusing. @Artifizer what do you think?

MikeFalcon77 commented on modules/types-registry/types-registry-sdk/src/api.rs:
```diff
@@ -0,0 +1,80 @@
+        &self,
+        ctx: &SecurityCtx,
+        entities: Vec<serde_json::Value>,
+    ) -> Result<Vec<GtsEntity>, TypesRegistryError>;
```
For batch-style APIs, it may be useful to return per-item results (ok/error) rather than all-or-nothing behavior, especially for any tooling

alizid10 replied:
done

MikeFalcon77 commented on modules/types-registry/types-registry-sdk/src/models.rs:
```diff
@@ -0,0 +1,311 @@
+}
+
+/// Type alias for dynamic GTS entities using `serde_json::Value` as content.
+pub type DynGtsEntity = GtsEntity<serde_json::Value>;
```
Probably we can simplify DX here with additional type alisases like:

pub struct TypeSchema(pub serde_json::Value);
pub struct InstanceObject(pub serde_json::Value);

pub type GtsTypeEntity = GtsEntity<TypeSchema>;
pub type GtsInstanceEntity = GtsEntity<InstanceObject>;

alizid10 replied:
added

alizid10 committed:
feat(types-registry-sdk): improve SDK models API and DX

- Remove serde derives from ListQuery (SDK guideline compliance)
- Add SegmentMatchScope enum for explicit filter matching behavior
- Remove unnecessary trait bounds from GtsEntity<C>
- Add RegisterResult for per-item batch operation results
- Add TypeSchema/InstanceObject newtypes with Deref/AsRef
- Add GtsTypeEntity/GtsInstanceEntity type aliases
- Clarify UUID v5 namespace documentation

Signed-off-by: Anton S <aliziddev@gmail.com>

alizid10 commented on modules/types-registry/types-registry-sdk/src/models.rs:
```diff
@@ -0,0 +1,311 @@
+
+    /// Sets the pattern filter.
+    #[must_use]
+    pub fn with_pattern(mut self, pattern: impl Into<String>) -> Self {
```
added

alizid10 committed:
chore: update gts-rust submodule to latest main

Signed-off-by: Anton S <aliziddev@gmail.com>
